File: tests/validation.test.ts
import { describe, test, expect } from 'bun:test';
import {
  isValidNumber,
  parseNumber,
  isValidUrl,
  parseStringList,
} from '../src/utils/validation.js';

describe('isValidNumber', () => {
  test('returns true for valid numbers', () => {
    expect(isValidNumber(42)).toBe(true);
    expect(isValidNumber(3.14)).toBe(true);
    expect(isValidNumber(0)).toBe(true);
    expect(isValidNumber(-10)).toBe(true);
  });

  test('returns true for numeric strings', () => {
    expect(isValidNumber('42')).toBe(true);
    expect(isValidNumber('3.14')).toBe(true);
    expect(isValidNumber('0')).toBe(true);
    expect(isValidNumber('-10')).toBe(true);
  });

  test('returns false for invalid values', () => {
    expect(isValidNumber('abc')).toBe(false);
    expect(isValidNumber('')).toBe(false);
    expect(isValidNumber(null)).toBe(false);
    expect(isValidNumber(undefined)).toBe(false);
    expect(isValidNumber({})).toBe(false);
    expect(isValidNumber([])).toBe(false);
    expect(isValidNumber(NaN)).toBe(false);
  });
});

describe('parseNumber', () => {
  test('parses valid numbers', () => {
    expect(parseNumber(42)).toBe(42);
    expect(parseNumber('42')).toBe(42);
    expect(parseNumber('3.14')).toBe(3.14);
  });

  test('returns undefined for invalid values', () => {
    expect(parseNumber('abc')).toBeUndefined();
    expect(parseNumber('')).toBeUndefined();
    expect(parseNumber(null)).toBeUndefined();
    expect(parseNumber(undefined)).toBeUndefined();
    expect(parseNumber({})).toBeUndefined();
  });
});

describe('isValidUrl', () => {
  test('returns true for valid HTTP/HTTPS URLs', () => {
    expect(isValidUrl('https://example.com')).toBe(true);
    expect(isValidUrl('http://localhost:3000')).toBe(true);
    expect(isValidUrl('https://example.com/path?query=value')).toBe(true);
  });

  test('returns false for non-HTTP/HTTPS URLs', () => {
    expect(isValidUrl('ftp://files.example.com')).toBe(false);
    expect(isValidUrl('file:///path/to/file')).toBe(false);
  });

  test('returns false for invalid URLs', () => {
    expect(isValidUrl('not-a-url')).toBe(false);
    expect(isValidUrl('')).toBe(false);
    expect(isValidUrl('example.com')).toBe(false);
    expect(isValidUrl('/path/to/resource')).toBe(false);
  });
});

describe('parseStringList', () => {
  test('parses comma-separated strings', () => {
    expect(parseStringList('a,b,c')).toEqual(['a', 'b', 'c']);
    expect(parseStringList('  a  ,  b  ,  c  ')).toEqual(['a', 'b', 'c']);
  });

  test('handles single values', () => {
    expect(parseStringList('single')).toEqual(['single']);
  });

  test('filters empty strings', () => {
    expect(parseStringList('a,,b,,c')).toEqual(['a', 'b', 'c']);
    expect(parseStringList(',,,')).toEqual([]);
  });

  test('returns undefined for non-strings', () => {
    expect(parseStringList(123)).toBeUndefined();
    expect(parseStringList(null)).toBeUndefined();
    expect(parseStringList(undefined)).toBeUndefined();
    expect(parseStringList(['a', 'b'])).toBeUndefined();
  });
});

File: internal/utils/validation_test.go
package utils

import (
	"reflect"
	"testing"
)

func TestParseNumber(t *testing.T) {
	cases := []struct {
		name   string
		input  string
		wantN  int
		wantOK bool
	}{
		{"positive integer", "42", 42, true},
		{"zero", "0", 0, true},
		{"negative integer", "-10", -10, true},
		{"empty string", "", 0, false},
		{"non-numeric", "abc", 0, false},
		{"float", "3.14", 0, false},
		{"whitespace only", "  ", 0, false},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			n, ok := ParseNumber(tc.input)
			if n != tc.wantN || ok != tc.wantOK {
				t.Errorf("ParseNumber(%q) = (%d, %t), want (%d, %t)", tc.input, n, ok, tc.wantN, tc.wantOK)
			}
		})
	}
}

func TestParseStringList(t *testing.T) {
	cases := []struct {
		name  string
		input string
		want  []string
	}{
		{"simple list", "a,b,c", []string{"a", "b", "c"}},
		{"with whitespace", "  a  ,  b  ", []string{"a", "b"}},
		{"single item", "single", []string{"single"}},
		{"empty segments", "a,,b", []string{"a", "b"}},
		{"all empty segments", ",,,", nil},
		{"empty string", "", nil},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := ParseStringList(tc.input)
			if !reflect.DeepEqual(got, tc.want) {
				t.Errorf("ParseStringList(%q) = %v, want %v", tc.input, got, tc.want)
			}
		})
	}
}

func TestIsValidURL(t *testing.T) {
	cases := []struct {
		name  string
		input string
		want  bool
	}{
		{"https URL", "https://example.com", true},
		{"http localhost", "http://localhost:3000", true},
		{"https with path and query", "https://example.com/path?query=value", true},
		{"ftp scheme", "ftp://files.example.com", false},
		{"no scheme", "not-a-url", false},
		{"empty string", "", false},
		{"bare domain", "example.com", false},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := IsValidURL(tc.input)
			if got != tc.want {
				t.Errorf("IsValidURL(%q) = %t, want %t", tc.input, got, tc.want)
			}
		})
	}
}

func TestIsValidSearchType(t *testing.T) {
	validTypes := []string{"auto", "neural", "keyword", "hybrid", "fast", "instant", "deep", "deep-lite", "deep-reasoning"}
	for _, st := range validTypes {
		t.Run("valid_"+st, func(t *testing.T) {
			if !IsValidSearchType(st) {
				t.Errorf("IsValidSearchType(%q) = false, want true", st)
			}
		})
	}

	invalidTypes := []string{"invalid", ""}
	for _, st := range invalidTypes {
		name := st
		if name == "" {
			name = "empty"
		}
		t.Run("invalid_"+name, func(t *testing.T) {
			if IsValidSearchType(st) {
				t.Errorf("IsValidSearchType(%q) = true, want false", st)
			}
		})
	}
}

func TestIsValidAnswerModel(t *testing.T) {
	cases := []struct {
		name  string
		input string
		want  bool
	}{
		{"exa", "exa", true},
		{"exa-pro", "exa-pro", true},
		{"gpt4", "gpt4", false},
		{"empty", "", false},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := IsValidAnswerModel(tc.input)
			if got != tc.want {
				t.Errorf("IsValidAnswerModel(%q) = %t, want %t", tc.input, got, tc.want)
			}
		})
	}
}

func TestIsValidResearchModel(t *testing.T) {
	cases := []struct {
		name  string
		input string
		want  bool
	}{
		{"fast", "fast", true},
		{"regular", "regular", true},
		{"pro", "pro", true},
		{"slow", "slow", false},
		{"empty", "", false},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := IsValidResearchModel(tc.input)
			if got != tc.want {
				t.Errorf("IsValidResearchModel(%q) = %t, want %t", tc.input, got, tc.want)
			}
		})
	}
}
